### PR TITLE
Validity and timelocks

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] reading properly validity from request body
- [x] add checks on boundary precedence and quantity
- [x] simplify how asset quantity is called in constructTransaction (in line with decodeTransaction)
- [x] slot check
- [x] make either validity bound optional
- [x] introduce before bound to transaction context and propagate this to node's tx
- [x] extend decodeTransaction to detect txValidityInterval (with integration testing)
- [x] introducing slot computation of scripts
- [x] script vs validity check
- [x] timelock minting test on   
- [x] add unit tests in MintBurnSpec

### Issue Number
adp-1193

### CODE REVIEW

**Comment by jonathanknowles** [comment]
bors try

**Comment by jonathanknowles** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 1]
```diff
@@ -4084,8 +4087,8 @@ data ApiMintData (n :: NetworkDiscriminant) = ApiMintData
         -- If no address is specified, then minted assets will be returned to
         -- the wallet as change, and change output addresses will be assigned
         -- automatically.
-    , amount
-        :: Quantity "assets" Natural
+    , quantity
+        :: Natural
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Types.hs:4091]: Hi @paweljakubas. Is this a breaking change to the API?

Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Api/Types.hs:4091]: well, yes. but new workflow is still experimental API ;-)

**Review by jonathanknowles — APPROVED** [review]

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs [hunk 2]
```diff
@@ -1232,6 +1263,58 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
...
+        \Validity interval 'unspecified'" $
+        \ctx -> runResourceT $ do
+
+        wa <- fixtureWallet ctx
+
+        let payload = Json [json|
+                { "withdrawal": "self"
+                , "validity_interval":
+                    { "invalid_before": "unspecified"
+                    , "invalid_hereafter": "unspecified"
```
Comment (piotr-iohk) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs]: The "unspecified" should be removed from the swagger then:
Screenshot from 2022-04-29 10-03-19

**Pull request merged**